### ISSUE TITLE
ios13 - can't debug on real device, mDns failure related

### ISSUE THREAD

**Issue by jaco-pixeldump** [flutter/flutter#41911]
having this issue on two different machines, they fail to establish debug connection, here the latest flutter run -v lines:

```
Checking for advertised Dart observatories...
[ +11 ms] Failed to establish a debug connection with ef0396dda2bb80d5a780985d0bd3f8c0008e2550 using mdns: SocketException:
Failed to
create datagram socket (OS Error: Address already in use, errno = 48), address = , port = 5353
[ +2 ms] mDNS lookup failed, attempting fallback to reading device log.
[ ] Waiting for observatory port.
```

it happens with all channels from stable to master
osx 10.14.1.6
ipad 5th gen ios 13.1.2
android studio 3.5.1
intellij 2019.2

**Comment by enricobenedos** [comment]
I've the same problem with `iOS 13.1.2` on iPad 2018 (6th gen), flutter `1.9.1+hotfix.4` and macOS `10.14.6`. It happens after the update of flutter from `1.9.1+hotfix.2` to `1.9.1+hotfix.4`. I needed to update flutter due to an iOS 13.1 bug #37699. 
This is the error:
```log
[  +24 ms] (lldb)     command script import "/tmp/034E69EC-D4C7-4696-B951-7D4E3540C0F4/fruitstrap_fe300c6fc96ac916a04068545991dce4733fcea4.py"
[   +3 ms] (lldb)     command script add -f fruitstrap_fe300c6fc96ac916a04068545991dce4733fcea4.connect_command connect
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_fe300c6fc96ac916a04068545991dce4733fcea4.run_command run
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_fe300c6fc96ac916a04068545991dce4733fcea4.autoexit_command autoexit
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_fe300c6fc96ac916a04068545991dce4733fcea4.safequit_command safequit
[        ] (lldb)     connect
[  +28 ms] (lldb)     run
[ +153 ms] success
[        ] (lldb)     safequit
[ +110 ms] Process 392 detached
[  +41 ms] Application launched on the device. Waiting for observatory port.
[   +3 ms] Checking for advertised Dart observatories...
[+5019 ms] mDNS lookup failed, attempting fallback to reading device log.
[        ] Waiting for observatory port.
```

**Comment by EricPreventa** [comment]
Also having this issue. Anyone else experience it? Any solution to it?

**Comment by James-A-White** [comment]
Yep... I'm also having this problem on an iPhone 11. I cannot launch from Visual Studio Code. The screen appears blank and I get the error:

`Failed to establish a debug connection with 00008030-000C44C422A0802E using mdns: SocketException: Failed to create datagram socket (OS Error: Address already in use, errno = 48), address = , port = 5353`

I can, however, quit the debug session in VS Code and run the app that was downloaded to the phone with no problems.

**Comment by jbnixdorf** [comment]
@enricobenedos: I'm not sure if we have the same issue as @jaco-pixeldump because we don't get the following lines:

```
[ +11 ms] Failed to establish a debug connection with ef0396dda2bb80d5a780985d0bd3f8c0008e2550 using mdns: SocketException:
Failed to create datagram socket (OS Error: Address already in use, errno = 48), address = , port = 5353
```

I already attached my current log to #41085. And I'm not getting a black screen - for me the app is starting fine, but flutter cannot connect.

**Comment by jaco-pixeldump** [comment]
just to point out to the title of the topic which is not related to the building phase, just to severe debug connection fail, it's still blocking and should be fixed asap.

**Comment by enricobenedos** [comment]
@jbnixdorf you are right, thank you for sharing the #41085 issue. 
But in my case the app does not start, my iPad shows blank screen and VSCode freeze at 
`[        ] Waiting for observatory port.`
As explained by you also in my case it is not sporadic but it always happens after the last updates.

My debug log was obtained using `flutter run -v` command

**Comment by jaco-pixeldump** [comment]
> My debug log was obtained using `flutter run -v` command

yes, app won't start due to debug conn fail, but It build and install on device, if you launch from device it starts  assuming there are no further app issues.

**Comment by DjangoOverloard** [comment]
It seems that I have a different type of error, it says something like
```
Failed to establish a debug connection with <My device> using mdns: OS Error:
Operation not supported on socket, errno = 102
```
Can these problems be related?

**Comment by zanderso** [comment]
@dnfield can you advise on the "Address already in use" failure in mdns?

**Comment by dnfield** [comment]
We should be setting up the socket with SO_REUSEADDR. It looks like @jonahwilliams added a patch to make this configurable.  Are we sure the tool is still passing in `reuseAddr` and `reusePort` as true?

**Comment by christopherfujino** [comment]
so this is happening because we're trying to open a second connection on the same port, I'm working on a fix for this https://github.com/flutter/flutter/pull/42026

**Comment by bemesquita** [comment]
Same with me. `flutter run -v` log:

```
[   +3 ms] Checking for advertised Dart observatories...
[+5020 ms] mDNS lookup failed, attempting fallback to reading device log.
[        ] Waiting for observatory port.
```

**Comment by christopherfujino** [comment]
Closed in #42026

**Comment by bemesquita** [comment]
@christopherfujino I pulled from master but still happening. Do I need to run any command to update?

**Comment by christopherfujino** [comment]
@bemesquita that PR specifically fixes the `OS Error: Address already in use, errno = 48` issue. I'm not seeing that in your verbose logs.

What this could be is that you haven't accepted a macOS pop-up dialog to allow incoming network connections from the `dart` binary. It will pop up if you haven't already accepted it.

**Comment by bemesquita** [comment]
@christopherfujino what pop-up dialog is that? sorry, I am new on flutter development.

**Comment by christopherfujino** [comment]
From your macOS firewall. It would look something like https://i.stack.imgur.com/5x54f.png

**Comment by bemesquita** [comment]
All good now @christopherfujino ! Thanks man!

**Comment by barees63** [comment]
I'm on the flutter dev channel and I'm still getting this error when attempting to run on physical device

`Failed to establish a debug connection with aef75...aa3 using mdns: OS Error: Operation not supported on socket, errno = 102`

Works fine on simulator and debugging the native side direct from XCode.
The device is running iOS 13.1.2 and I'm on MacOS Catalina, is this a separate issue?

**Comment by christopherfujino** [comment]
@barees63 I haven't seen that error message before; yes, I believe that is new. Could you please file a new issue with full `flutter doctor -v` and `flutter run -v` logs and then tag me on it? Thanks.

**Comment by jaco-pixeldump** [comment]
sorry to report, nothing changed yet, same errors.
flutter master ch. 1.10.15-pre78

error still persists on both mohave/chsh and catalina/zsh
they both runs on different hw, pretty fresh install, latest xcode, no firewall, file permission granted.

**Comment by christopherfujino** [comment]
Hmm, that's definitely past my fix, @jaco-pixeldump. Could you give me your full verbose `flutter doctor -v` logs? And I'm assuming you're getting the same error message as in the original post, `Failed to
create datagram socket (OS Error: Address already in use, errno = 48), address = , port = 5353`?

**Comment by christopherfujino** [comment]
@jaco-pixeldump oh, also, can you try `killall -9 iproxy`? Part of my fix was to stop leaking these processes, but you could still have them running.

**Comment by jaco-pixeldump** [comment]
@christopherfujino, thank you

yes error is exactly what I've reported.
tried killall, no processes were found.

attached full log, hope this help to resolve.

* error is the same with or without apple provisioning check.

[build.log](https://github.com/flutter/flutter/files/3718837/build.log)

**Comment by christopherfujino** [comment]
Can you also share the output of `/usr/libexec/ApplicationFirewall/socketfilterfw --listapps`? This application will show you all the whitelist/blacklist entries in your firewall. In particular, I'm interested in any entries related to `flutter/bin/cache/dart-sdk/bin/dart`, and if your firewall is blocking them.

If they're being blocked by your firewall, you can unblock it in the UI by going:

System Preferences -> Security & Privacy -> Firewall -> Firewall Options -> then "Allow incoming connections" for Dart

**Comment by jaco-pixeldump** [comment]
/usr/libexec/ApplicationFirewall/socketfilterfw --listapps
produces empty result.
fw is disabled, so no options to check

**Comment by christopherfujino** [comment]
hmm....

**Comment by jaco-pixeldump** [comment]
I've also uninstalled chrome and Firefox just to be sure there is no mDns interferences.
my network setting is static, no proxy.
Tomorrow I'll try on second machine (Catalina), will report if there is some noticeable differences.

**Comment by christopherfujino** [comment]
@jaco-pixeldump I noticed in the log that you attached looks slightly different from the original post. I'm seeing:

```
[  +30 ms] Application launched on the device. Waiting for observatory port.
[   +3 ms] Checking for advertised Dart observatories...
[   +2 ms] mDNS lookup failed, attempting fallback to reading device log.
```

Which I *think* means that the original "address already in use" bug has been fixed...and you're running into another one, which is perhaps iPad specific. I'll continue to investigate...

**Comment by christopherfujino** [comment]
@jaco-pixeldump @enricobenedos Ok, I have confirmed that there is a problem with mDNS on an iPad. I have created a [new issue](https://github.com/flutter/flutter/issues/42545) to track this, as this issue, about "address already in use" has been fixed. Please follow there!

**Comment by jaco-pixeldump** [comment]
still blocking with same errors on iPhone 7.

**Comment by christopherfujino** [comment]
@jaco-pixeldump can you provide full error logs from your iPhone 7? As there are multiple issues going on, I would like to narrow down which issues affect which devices. Thank you!

**Comment by jaco-pixeldump** [comment]
@christopherfujino 
here the logs, thank you.

[run.log](https://github.com/flutter/flutter/files/3730891/run.log)
[doctor.log](https://github.com/flutter/flutter/files/3730893/doctor.log)

**Comment by jaweii** [comment]
same problem here.

**Comment by TheKashe** [comment]
Having the same issue here, iPhone 6s on ios 12.4.1. It worked fine until yesterday when I had to clean everything and rebuild to be able to do development deploy to other user's iphone through apple configurator.

Since then, I can't deploy through VSC or flutter run:

    [ +118 ms] Process 454 detached
    [  +43 ms] Application launched on the device. Waiting for observatory port.
    [   +7 ms] Checking for advertised Dart observatories...
    [+5031 ms] mDNS lookup failed, attempting fallback to reading device log.
    [        ] Waiting for observatory port.

Clean, rebuild, nothing works. Xcode deployment works, but doesn't fix the issue.

**Comment by christopherfujino** [comment]
@jaweii can you please provide full `flutter doctor -v` and `flutter run -v` logs? Thanks

**Comment by DjangoOverloard** [comment]
@christopherfujino, I've commented about this issue about 12 days ago and nothing changed pretty much, maybe that's going to help you track the problem down. I also have the iPhone 7 and the same thing happens with my iPad Pro (11 inch 2018). I did not upgrade to Catalina yet. 

Here is the part of the log

```
[  +51 ms] Application launched on the device. Waiting for observatory port.
[   +3 ms] Checking for advertised Dart observatories...
[  +11 ms] Failed to establish a debug connection with <device name> using mdns: OS
Error:
           Operation not supported on socket, errno = 102
[   +2 ms] mDNS lookup failed, attempting fallback to reading device log.
[        ] Waiting for observatory port.
```

**Comment by jaco-pixeldump** [comment]
this issue is still there with same severe fail.
Please reopen and fix for real asap, thanks.

**Comment by jaweii** [comment]
@christopherfujino 

> Doctor summary (to see all details, run flutter doctor -v):
> [✓] Flutter (Channel stable, v1.9.1+hotfix.6, on Mac OS X 10.14.5 18F203, locale zh-Hans-CN)
>  
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
> [✓] Xcode - develop for iOS and macOS (Xcode 11.1)
> [!] Android Studio (version 3.4)
>     ✗ Flutter plugin not installed; this adds Flutter specific functionality.
>     ✗ Dart plugin not installed; this adds Dart specific functionality.
> [✓] Connected device (1 available)
> 
> ! Doctor found issues in 1 category.

**Comment by edTheGuy00** [comment]
I'm also having this issue only on my physical device. It started since updating to iOS 13 and Xcode 11. it is able to install and open the app just fine, but unable to connect at all.

`Failed to establish a debug connection with c07............... using mdns: OS Error: Operation not supported on socket, errno = 102`

**Comment by synax** [comment]
Also experiencing the same/similar issue with: 

`flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.9.1+hotfix.6, on Mac OS X 10.15 19A602, locale en-CH)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.39.2)
[✓] Connected device (3 available)`

`flutter run -v
....
+245 ms] [ 52%] CreatingStagingDirectory
[        ] [ 57%] ExtractingPackage
[        ] [ 60%] InspectingPackage
[  +16 ms] [ 60%] TakingInstallLock
[  +61 ms] [ 65%] PreflightingApplication
[  +58 ms] [ 65%] InstallingEmbeddedProfile
[  +92 ms] [ 70%] VerifyingApplication
[ +249 ms] [ 75%] CreatingContainer
[   +5 ms] [ 80%] InstallingApplication
[   +4 ms] [ 85%] PostflightingApplication
[   +3 ms] [ 90%] SandboxingApplication
[  +16 ms] [ 95%] GeneratingApplicationMap
[ +174 ms] [100%] Installed package build/ios/iphoneos/Runner.app
[ +278 ms] ------ Debug phase ------
[        ] Starting debug of a492df5ee130e1661d21fa53e7eb5f4852573ac7 (D221AP, iPhone X, iphoneos, arm64) a.k.a. 'synax_mobile' connected through USB...
[+1611 ms] [  0%] Looking up developer disk image
[  +16 ms] [ 95%] Developer disk image mounted successfully
[+1091 ms] [100%] Connecting to remote debug server
[        ] -------------------------
[ +249 ms] (lldb) command source -s 0 '/tmp/2F8B857F-455D-4516-976B-F149B208CA3B/fruitstrap-lldb-prep-cmds-a492df5ee130e1661d21fa53e7eb5f4852573ac7'
[        ] Executing commands in '/tmp/2F8B857F-455D-4516-976B-F149B208CA3B/fruitstrap-lldb-prep-cmds-a492df5ee130e1661d21fa53e7eb5f4852573ac7'.
[        ] (lldb)     platform select remote-ios --sysroot '/Users/stefan/Library/Developer/Xcode/iOS DeviceSupport/13.1.3 (17A878)/Symbols'
[        ]   Platform: remote-ios
[        ]  Connected: no
[        ]   SDK Path: "/Users/stefan/Library/Developer/Xcode/iOS DeviceSupport/13.1.3 (17A878)/Symbols"
[        ] (lldb)     target create "/Users/stefan/GitHub/project-wayland/flutter-firebase-app/build/ios/iphoneos/Runner.app"
[+5299 ms] Current executable set to '/Users/stefan/GitHub/project-wayland/flutter-firebase-app/build/ios/iphoneos/Runner.app' (arm64).
[        ] (lldb)     script fruitstrap_device_app="/private/var/containers/Bundle/Application/7A083135-AF5C-45CA-8546-C699A49ECF16/Runner.app"
[ +419 ms] (lldb)     script fruitstrap_connect_url="connect://127.0.0.1:55335"
[        ] (lldb)     script fruitstrap_output_path=""
[        ] (lldb)     script fruitstrap_error_path=""
[        ] (lldb)     target modules search-paths add /usr "/Users/stefan/Library/Developer/Xcode/iOS DeviceSupport/13.1.3 (17A878)/Symbols/usr" /System
"/Users/stefan/Library/Developer/Xcode/iOS DeviceSupport/13.1.3 (17A878)/Symbols/System" "/private/var/containers/Bundle/Application/7A083135-AF5C-45CA-8546-C699A49ECF16"
"/Users/stefan/GitHub/project-wayland/flutter-firebase-app/build/ios/iphoneos" "/var/containers/Bundle/Application/7A083135-AF5C-45CA-8546-C699A49ECF16"
"/Users/stefan/GitHub/project-wayland/flutter-firebase-app/build/ios/iphoneos" /Developer "/Users/stefan/Library/Developer/Xcode/iOS DeviceSupport/13.1.3 (17A878)/Symbols/Developer"
[  +27 ms] (lldb)     command script import "/tmp/2F8B857F-455D-4516-976B-F149B208CA3B/fruitstrap_a492df5ee130e1661d21fa53e7eb5f4852573ac7.py"
[   +5 ms] (lldb)     command script add -f fruitstrap_a492df5ee130e1661d21fa53e7eb5f4852573ac7.connect_command connect
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_a492df5ee130e1661d21fa53e7eb5f4852573ac7.run_command run
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_a492df5ee130e1661d21fa53e7eb5f4852573ac7.autoexit_command autoexit
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_a492df5ee130e1661d21fa53e7eb5f4852573ac7.safequit_command safequit
[        ] (lldb)     connect
[  +36 ms] (lldb)     run
[ +210 ms] success
[        ] (lldb)     safequit
[ +108 ms] Process 8583 detached
[  +68 ms] Application launched on the device. Waiting for observatory port.
[   +4 ms] Checking for advertised Dart observatories...
[+5022 ms] mDNS lookup failed, attempting fallback to reading device log.
[        ] Waiting for observatory port.
`

**Comment by jerryzhoujw** [comment]
I got the same error of `mDNS lookup failed, attempting fallback to reading device log `

**Comment by jaco-pixeldump** [comment]
no real news, no changes from more than 1 month, embarrassing.

The only chance I get to print out some log is via xcode on its console.

**Comment by CodeMasterAlexey** [comment]
Got the Error still after downgrading to 1.9 Hotfix +4 or even further like 1.8 ..
Any News? Cant really Log because the Console is blocked with the Installing text..

**Comment by deche** [comment]
I'm having the same problem after upgrading flutter to the latest 1.9.1+hotfix.6 

Here's last few lines of flutter run -v:

[        ] (lldb)     command script add -s asynchronous -f fruitstrap_c0725b51c8cd7d948abbe65aa588a8f73e378e0e.run_command run
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_c0725b51c8cd7d948abbe65aa588a8f73e378e0e.autoexit_command autoexit
[        ] (lldb)     command script add -s asynchronous -f fruitstrap_c0725b51c8cd7d948abbe65aa588a8f73e378e0e.safequit_command safequit
[        ] (lldb)     connect
[  +38 ms] (lldb)     run
[ +230 ms] success
[        ] (lldb)     safequit
[ +114 ms] Process 11460 detached
[  +50 ms] Application launched on the device. Waiting for observatory port.
[   +3 ms] Checking for advertised Dart observatories...
[+5019 ms] mDNS lookup failed, attempting fallback to reading device log.
[        ] Waiting for observatory port.

**Comment by TheFedex87** [comment]
@deche any news about this problem? I'm stuck with the same messages. App is installed but it is stuck with a blank screen after it is run. :(

**Comment by deche** [comment]
I've found the solution to my problem. Just upgrade your iphone to the latest ios. In my case it's 13.2.3
Hope it helps

**Comment by TheFedex87** [comment]
@deche already updated at the version 13.2.3 but still stuck into the same problem :(

**Comment by jinsong5** [comment]
mDNS lookup failed, attempting fallback to reading device log

**Comment by perepechin** [comment]
> @deche already updated at the version 13.2.3 but still stuck with the same problem :(

I finally found a solution which works for me at least in VS Code. Please check out this comment:
https://github.com/flutter/flutter/issues/46698#issuecomment-565653904

In my case on my iPhone I have iOS 13.3 and I am using Xcode 11.3.
